File: flutter/dominion_randomizer/lib/models/cost.dart
import 'dart:math';

import 'package:quiver/core.dart';

enum CostType {
  treasure2OrLess,
  treasure3,
  treasure4,
  treasure5,
  treasure6,
  treasure7,
  treasure8OrMore,
}

class Cost {
  final int treasure;
  final int potion;
  final int debt;

  Cost(
    this.treasure,
    this.potion,
    this.debt,
  );

  CostType getType() {
    final costs = [
      CostType.treasure2OrLess,
      CostType.treasure2OrLess,
      CostType.treasure2OrLess,
      CostType.treasure3,
      CostType.treasure4,
      CostType.treasure5,
      CostType.treasure6,
      CostType.treasure7,
      CostType.treasure8OrMore,
    ];
    return costs[min(treasure, 8)];
  }

  Cost.fromJson(Map<String, dynamic>? json)
    : treasure = json?['treasure'] as int? ?? 0,
      potion = json?['potion'] as int? ?? 0,
      debt = json?['debt'] as int? ?? 0;

  @override
  bool operator ==(o) => o is Cost && o.treasure == treasure && o.potion == potion && o.debt == debt;

  @override
  int get hashCode => hash3(treasure, potion, debt);
}

File: src/dominion/cost.ts
import { CostType } from "./cost-type"

export class Cost {
  constructor(
    readonly treasure: number,
    readonly potion: number,
    readonly debt: number) {
  }

  getType(): CostType {
    const costs: CostType[] = [
      CostType.TREASURE_2_OR_LESS,
      CostType.TREASURE_2_OR_LESS,
      CostType.TREASURE_2_OR_LESS,
      CostType.TREASURE_3,
      CostType.TREASURE_4,
      CostType.TREASURE_5,
      CostType.TREASURE_6,
      CostType.TREASURE_7,
      CostType.TREASURE_8_OR_MORE,
    ];
    return costs[Math.min(this.treasure, 8)];
  }

  public static fromJson(json: any) {
    return new Cost(
      json["treasure"] || 0,
      json["potion"] || 0,
      json["debt"] || 0);
  }
}